keps/sig-cli/0031-kustomize-integration.md

```diff
@@ -0,0 +1,230 @@
+
+## Kustomize Introduction
+
+Kustomize has following subcommands:
+- build
+- edit, edit also has subcommands
+   - Set 
+       - imagetag
+       - namespace
+       - nameprefix
+   - Add
+       - base
+       - resource
+       - patch
+       - label
+       - annotation
+       - configmap
+- config
+- version
+- help
+
+`edit` and `build` are most commonly used subcommands. 
+
+`edit` is to modify the fields in `kustomization.yaml`. A `kustomization.yaml` includes configurations that are consumed by Kustomize. Here is an example of `kustomization.yaml` file.
+
+`build` is to perform a set of pre-processing transformations on the resources inside one kustomization. Those transformations include:
+- Get objects from the base
+- Apply patches
+- Add name prefix to all resources
+- Add common label and annotation to all resources
+- Replace imageTag is specified
+- Update objects’ names where they are referenced
+- Resolve variables and substitute them
+
+```
+namePrefix: alices-
+
+commonAnnotations:
+ oncallPager: 800-555-1212
+
+configMapGenerator:
+- name: myJavaServerEnvVars
+ literals:   
+ - JAVA_HOME=/opt/java/jdk
+ - JAVA_TOOL_OPTIONS=-agentlib:hprof
+
+secretGenerator:
+- name: app-sec
+ commands:
+   username: "echo admin"
+   password: "echo secret"
+```
+The build output of this sample kustomizaiton.yaml file is
+```
+apiVersion: v1
+data:
+ JAVA_HOME: /opt/java/jdk
+ JAVA_TOOL_OPTIONS: -agentlib:hprof
+kind: ConfigMap
+metadata:
+ annotations:
+   oncallPager: 800-555-1212
+ name: alices-myJavaServerEnvVars-7bc9c27cmf
+---
+apiVersion: v1
+data:
+ password: c2VjcmV0Cg==
+ username: YWRtaW4K
+kind: Secret
+metadata:
+ annotations:
+   oncallPager: 800-555-1212
+ name: alices-app-sec-c7c5tbh526
+type: Opaque
+```
+
+## Proposal
+
+### UX
+
+When apply, get or delete is run on a directory, check if it contains a kustomization.yaml file. If there is, apply, get or delete the output of kustomize build. Kubectl behaves the same as current for directories without kustomization.yaml.
+
+#### apply
+The command visible to users is
+```
+kubectl apply -f <dir>
+```
+To view the objects in a kustomization without applying them to the cluster
+```
+kubectl apply -f <dir> --dry-run -o yaml|json
+```
+
+#### get
+The command visible to users is
+```
+kubectl get -f <dir>
+```
+To get the detailed objects in a kustomization
+```
+kubectl get -f <dir> --dry-run -o yaml|json
+```
+
+#### delete
+The command visible to users is
+```
+kubectl delete -f <dir>
+```
+
+### Implementation Details/Notes/Constraints
+
+To enable kustomize in kubectl, the function `FilenameParam` inside Builder type will be updated to recognize kustomization directories. The Builder will expand the sources in a kustomization directory and pass them to a subcommand.
+
+ Since kustomization directories themselves have a recursive structure, `-R` will be ignored on those directories. Allowing recursive visit to the same files will lead to duplicate resources. 
```

Comment:
Why ignored rather than error? What do we think users would want?

Reply:
Users may `apply -R` on both a kustomization directory and a non kustomization directory. They would expect this to be successfully applied. If we throw an error, this use case will not be supported.

Reply:
Makes sense. Please ensure it's clear in the documentation.

Reply:
Sure. We will write the documentation after the code PR is merged.